core_impl/src/test/java/io/opencensus/trace/internal/ConcurrentIntrusiveListTest.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2017, Google Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.trace.internal;
+
+import static com.google.common.truth.Truth.assertThat;
+
+import io.opencensus.trace.internal.ConcurrentIntrusiveList.Element;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Unit tests for {@link ConcurrentIntrusiveList}. */
+@RunWith(JUnit4.class)
+public class ConcurrentIntrusiveListTest {
+  private final ConcurrentIntrusiveList<FakeElement> intrusiveList =
+      new ConcurrentIntrusiveList<FakeElement>();
+  @Rule public final ExpectedException exception = ExpectedException.none();
+
+  @Test
+  public void emptyList() {
+    assertThat(intrusiveList.isEmpty()).isTrue();
+    assertThat(intrusiveList.size()).isEqualTo(0);
+    assertThat(intrusiveList.getAll().isEmpty()).isTrue();
+  }
+
+  @Test
+  public void addRemoveAdd_SameElement() {
+    FakeElement element = new FakeElement();
+    intrusiveList.addElement(element);
+    assertThat(intrusiveList.isEmpty()).isFalse();
+    assertThat(intrusiveList.size()).isEqualTo(1);
+    intrusiveList.removeElement(element);
+    assertThat(intrusiveList.isEmpty()).isTrue();
+    assertThat(intrusiveList.size()).isEqualTo(0);
+    intrusiveList.addElement(element);
+    assertThat(intrusiveList.isEmpty()).isFalse();
+    assertThat(intrusiveList.size()).isEqualTo(1);
+  }
+
+  @Test
+  public void addAndRemoveElements() {
+    FakeElement element1 = new FakeElement();
+    FakeElement element2 = new FakeElement();
+    FakeElement element3 = new FakeElement();
+    intrusiveList.addElement(element1);
+    intrusiveList.addElement(element2);
+    intrusiveList.addElement(element3);
+    assertThat(intrusiveList.size()).isEqualTo(3);
+    assertThat(intrusiveList.getAll().contains(element1)).isTrue();
+    assertThat(intrusiveList.getAll().contains(element2)).isTrue();
+    assertThat(intrusiveList.getAll().contains(element3)).isTrue();
```

Comment:
Suggestion: You could use containsExactly to also check the order, i.e.,

    assertThat(intrusiveList.getAll()).containsExactly(element1, element2, element3).inOrder();